File: bot.js
const { Telegraf, Markup, session } = require('telegraf');
const { message } = require('telegraf/filters');
const axios = require('axios');
const { v4: uuidv4 } = require('uuid');
const express = require('express');

// =========================
// Configuration & Constants
// =========================

const TELEGRAM_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const WEBHOOK_URL = process.env.RENDER_EXTERNAL_URL || process.env.WEBHOOK_URL;
const PORT = process.env.PORT || 8000;
const DEV_HANDLE = "@yaplol";

// Channel Configuration
const REQUIRED_CHANNELS = [
    { 
        id: "-1002628211220", 
        name: "MAIN CHANNEL 📢", 
        link: "https://t.me/c/2628211220/1" 
    },
    { 
        id: "@pytimebruh", 
        name: "BACKUP CHANNEL", 
        link: "https://t.me/pytimebruh" 
    },
];

// Validation
if (!TELEGRAM_TOKEN) {
    console.error("❌ FATAL: TELEGRAM_BOT_TOKEN environment variable not set");
    process.exit(1);
}

// =========================
// Bot Initialization
// =========================

const bot = new Telegraf(TELEGRAM_TOKEN);
const app = express();

// Middleware Configuration
app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true }));

// Application State Management
const applicationState = {
    initialized: false,
    webhookVerified: false,
    error: null,
    startupTime: new Date().toISOString(),
    statistics: {
        totalRequests: 0,
        successfulResets: 0,
        failedResets: 0
    }
};

// =========================
// Utility Functions
// =========================

/**
 * Membership Verification Service
 */
class MembershipService {
    static async verifyUserMembership(userId, ctx) {
        try {
            const verificationPromises = REQUIRED_CHANNELS.map(channel => 
                this.checkSingleChannel(userId, channel.id, ctx)
            );
            
            const results = await Promise.allSettled(verificationPromises);
            return results.map(result => 
                result.status === 'fulfilled' ? result.value : false
            );
        } catch (error) {
            console.error(`Membership verification failed for user ${userId}:`, error);
            return REQUIRED_CHANNELS.map(() => false);
        }
    }

    static async checkSingleChannel(userId, channelId, ctx) {
        try {
            const member = await ctx.telegram.getChatMember(channelId, userId);
            return ['member', 'administrator', 'creator'].includes(member.status);
        } catch (error) {
            if (error.description?.includes('user not found')) {
                return false;
            }
            console.warn(`Channel check failed for ${channelId}:`, error.message);
            return false;
        }
    }

    static generateVerificationKeyboard() {
        const buttons = REQUIRED_CHANNELS.map(channel => 
            [Markup.button.url(channel.name, channel.link)]
        );
        buttons.push([Markup.button.callback('✅ Verify Membership', 'verify_membership')]);
        
        return Markup.inlineKeyboard(buttons);
    }
}

/**
 * Instagram Reset Service
 */
class InstagramResetService {
    constructor() {
        this.httpClient = axios.create({
            timeout: 15000,
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
        });
    }

    async executeResetMethods(target) {
        const methods = [
            this.methodEmailReset.bind(this),
            this.methodAdvancedReset.bind(this),
            this.methodWebReset.bind(this)
        ];

        try {
            const results = await Promise.allSettled(
                methods.map(method => method(target))
            );
            
            const successful = results.some(result => 
                result.status === 'fulfilled' && result.value === true
            );
            
            this.updateStatistics(successful);
            return successful;
        } catch (error) {
            console.error(`Reset execution error for ${target}:`, error.message);
            this.updateStatistics(false);
            return false;
        }
    }

    async methodEmailReset(target) {
        try {
            const response = await this.httpClient.post(
                'https://www.instagram.com/accounts/account_recovery_send_ajax/',
                `email_or_username=${encodeURIComponent(target)}`,
                { headers: { 'X-Requested-With': 'XMLHttpRequest' } }
            );
            return response.status === 200 && response.data.includes('email_sent');
        } catch {
            return false;
        }
    }

    async methodAdvancedReset(target) {
        try {
            const profileResponse = await this.httpClient.get(
                `https://www.instagram.com/api/v1/users/web_profile_info/?username=${target}`,
                { headers: { 'X-IG-App-ID': '936619743392459' } }
            );
            
            const userId = profileResponse.data.data.user.id;
            const deviceId = uuidv4();
            
            const resetResponse = await this.httpClient.post(
                'https://i.instagram.com/api/v1/accounts/send_password_reset/',
                `user_id=${userId}&device_id=${deviceId}`,
                {
                    headers: {
                        'User-Agent': 'Instagram 6.12.1 Android (30/11; 480dpi; 1080x2004; HONOR; ANY-LX2; HNANY-Q1; qcom; ar_EG_#u-nu-arab)',
                        'Content-Type': 'application/x-www-form-urlencoded'
                    }
                }
            );
            
            return resetResponse.data.includes('obfuscated_email');
        } catch {
            return false;
        }
    }

    async methodWebReset(target) {
        try {
            const response = await this.httpClient.post(
                'https://www.instagram.com/api/v1/web/accounts/account_recovery_send_ajax/',
                `email_or_username=${encodeURIComponent(target)}&flow=fxcal`,
                {
                    headers: {
                        'x-csrftoken': 'missing',
                        'x-ig-app-id': '936619743392459',
                        'x-requested-with': 'XMLHttpRequest',
                        'Content-Type': 'application/x-www-form-urlencoded'
                    }
                }
            );
            return response.data.status === 'ok';
        } catch {
            return false;
        }
    }

    updateStatistics(success) {
        applicationState.statistics.totalRequests++;
        if (success) {
            applicationState.statistics.successfulResets++;
        } else {
            applicationState.statistics.failedResets++;
        }
    }
}

// =========================
// Message Handlers
// =========================

/**
 * Welcome Message Handler
 */
async function displayWelcomeMessage(ctx) {
    const welcomeText = `
🤖 *Instagram Reset Bot*

*Commands:*
/rst - Single account reset
/blk - Bulk reset (max 5 accounts)

*Developer:* ${DEV_HANDLE}

Use the commands below to get started.
    `.trim();

    await ctx.replyWithMarkdown(welcomeText);
}

/**
 * Membership Verification Handler
 */
async function handleMembershipVerification(ctx, showStatus = false) {
    try {
        const statuses = await MembershipService.verifyUserMembership(ctx.from.id, ctx);
        const allJoined = statuses.every(status => status);

        if (allJoined) {
            return true;
        }

        let message = `🔒 *Access Required*\n\n`;
        
        if (showStatus) {
            message += `*Membership Status:*\n`;
            REQUIRED_CHANNELS.forEach((channel, index) => {
                const icon = statuses[index] ? '✅' : '❌';
                message += `${icon} ${channel.name}\n`;
            });
            message += `\nPlease join all required channels and verify again.`;
        } else {
            message += `To use this bot, please join our official channels.\n\n`;
            message += `Join all channels below and click *Verify Membership*.`;
        }

        await ctx.replyWithMarkdown(
            message,
            MembershipService.generateVerificationKeyboard()
        );
        return false;

    } catch (error) {
        console.error('Membership verification error:', error);
        await ctx.reply('⚠️ System temporarily unavailable. Please try again shortly.');
        return false;
    }
}

/**
 * Command Handlers
 */
async function handleStartCommand(ctx) {
    if (ctx.chat.type !== 'private') {
        return;
    }

    const isMember = await handleMembershipVerification(ctx, false);
    if (isMember) {
        await displayWelcomeMessage(ctx);
    }
}

async function handleVerificationCallback(ctx) {
    try {
        await ctx.answerCbQuery();
        const isMember = await handleMembershipVerification(ctx, true);
        
        if (isMember) {
            await ctx.editMessageText(
                '✅ *Verification Successful!*\n\nYou can now use all bot features.',
                { parse_mode: 'Markdown' }
            );
            await displayWelcomeMessage(ctx);
        }
    } catch (error) {
        console.error('Verification callback error:', error);
        await ctx.answerCbQuery('❌ Verification failed. Please try again.', { show_alert: true });
    }
}

async function handleHelpCommand(ctx) {
    const helpText = `
📖 *Command Guide*

*/rst username* - Reset single account
Example: \`/rst example_user\`

*/blk user1 user2* - Reset multiple accounts (max 5)
Example: \`/blk user1 user2 user3\`

*Developer:* ${DEV_HANDLE}
    `.trim();

    await ctx.replyWithMarkdown(helpText);
}

/**
 * Reset Command Handlers
 */
async function handleSingleReset(ctx) {
    try {
        const args = ctx.message.text.split(' ').slice(1);
        if (args.length === 0) {
            await ctx.replyWithMarkdown(
                '❌ *Usage:* `

File: bot.py
import os
import uuid
import string
import random
import requests
import logging
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, CallbackQueryHandler, ContextTypes, MessageHandler, filters

# Bot configuration
BOT_TOKEN = "8256075803:AAEBqIpIC514IcY-9HptJyAJA4XIdP8CDog"

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

class InstagramResetBot:
    def __init__(self):
        self.user_states = {}

    async def send_password_reset(self, target: str) -> dict:
        """Send password reset request to Instagram"""
        try:
            if '@' in target:
                data = {
                    '_csrftoken': ''.join(random.choices(string.ascii_letters + string.digits, k=32)),
                    'user_email': target,
                    'guid': str(uuid.uuid4()),
                    'device_id': str(uuid.uuid4())
                }
            else: 
                data = {
                    '_csrftoken': ''.join(random.choices(string.ascii_letters + string.digits, k=32)),
                    'username': target,
                    'guid': str(uuid.uuid4()),
                    'device_id': str(uuid.uuid4())
                }
            
            user_agent_parts = [
                ''.join(random.choices(string.ascii_lowercase + string.digits, k=16)),
                ''.join(random.choices(string.ascii_lowercase + string.digits, k=16)),
                ''.join(random.choices(string.ascii_lowercase + string.digits, k=16)),
                ''.join(random.choices(string.ascii_lowercase + string.digits, k=16))
            ]
            
            headers = {
                'user-agent': f"Instagram 150.0.0.0.000 Android (29/10; 300dpi; 720x1440; {user_agent_parts[0]}/{user_agent_parts[1]}; {user_agent_parts[2]}; {user_agent_parts[3]}; en_GB;)"
            }
            
            response = requests.post(
                'https://i.instagram.com/api/v1/accounts/send_password_reset/',
                headers=headers,
                data=data,
                timeout=30
            )
            
            if 'obfuscated_email' in response.text or 'obfuscated_phone' in response.text:
                return {"success": True, "message": f"✅ Reset sent to `{target}`"}
            else:
                return {"success": False, "message": f"❌ Failed: `{target}`"}
                
        except Exception as e:
            return {"success": False, "message": f"❌ Error: `{target}`"}

    def create_start_keyboard(self) -> InlineKeyboardMarkup:
        """Create initial start keyboard"""
        keyboard = [
            [InlineKeyboardButton("🔓 Reset", callback_data="start_reset")]
        ]
        return InlineKeyboardMarkup(keyboard)

    def create_reset_again_keyboard(self) -> InlineKeyboardMarkup:
        """Create reset again keyboard after completion"""
        keyboard = [
            [InlineKeyboardButton("🔄 Reset Another Account", callback_data="reset_again")]
        ]
        return InlineKeyboardMarkup(keyboard)

    async def start_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        """Send welcome message"""
        welcome_text = """🔓 Reset your Instagram account password.
        
        by @yaplol"""
        
        keyboard = self.create_start_keyboard()
        
        await update.message.reply_text(
            welcome_text, 
            reply_markup=keyboard,
            parse_mode='Markdown'
        )

    async def button_callback(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        """Handle button callbacks"""
        query = update.callback_query
        await query.answer()
        
        user_id = query.from_user.id
        
        if query.data == "start_reset" or query.data == "reset_again":
            await self.show_reset_input(update, context)

    async def show_reset_input(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Show reset input instructions"""
        user_id = update.callback_query.from_user.id
        self.user_states[user_id] = "waiting_target"
        
        text = """Send username, email, or phone number:"""
        
        await update.callback_query.edit_message_text(text)

    async def handle_text_message(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        """Handle text messages"""
        user_id = update.effective_user.id
        state = self.user_states.get(user_id)
        
        if state == "waiting_target":
            await self.process_reset(update, context)
        else:
            keyboard = self.create_start_keyboard()
            await update.message.reply_text(
                "Use the button below:",
                reply_markup=keyboard
            )

    async def process_reset(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Process account reset"""
        user_id = update.effective_user.id
        target = update.message.text.strip()
        
        self.user_states[user_id] = None
        
        processing_msg = await update.message.reply_text(f"🔄 Processing `{target}`...", parse_mode='Markdown')
        
        result = await self.send_password_reset(target)
        keyboard = self.create_reset_again_keyboard()
        
        await processing_msg.edit_text(
            result['message'],
            parse_mode='Markdown',
            reply_markup=keyboard
        )

    async def error_handler(self, update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
        """Log errors"""
        logger.error(f"Exception: {context.error}")

def main() -> None:
    """Start the bot"""
    print("🤖 Starting Instagram Reset Bot...")
    
    bot = InstagramResetBot()
    application = Application.builder().token(BOT_TOKEN).build()
    
    application.add_handler(CommandHandler("start", bot.start_command))
    application.add_handler(CallbackQueryHandler(bot.button_callback))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, bot.handle_text_message))
    application.add_error_handler(bot.error_handler)
    
    print("🚀 Bot running...")
    application.run_polling(allowed_updates=Update.ALL_TYPES)

if __name__ == "__main__":
    main()
